R/query_library.R

```diff
@@ -0,0 +1,238 @@
+#' \code{get_query("emissions", "co2_emissions")}.
+#' @param query_file (string) The path to a YAML file to read if not NULL otherwise
+#' the package query file will be used.
+#' @param ... (string) Strings specifiying the path to the query to lookup.
+#' @return (string) The raw query from the query file with units set as an attribute
+#' if specified.
+get_query <- function(..., query_file = NULL) {
+    if(is.null(query_file)) {
+        queries = PACKAGE_QUERIES
+    } else {
+        queries = read_yaml(query_file)
+    }
+    query_path <- list(...)
+    query <- queries
+    for(path in query_path) {
+        query <- query[[path]]
+        if(is.null(query)) {
+            stop(paste0("Could not find query: ", query_path))
+        }
+    }
+    if(length(query) == 0) {
+        stop(paste0("Could not find query: ", query_path))
+    }
+    query_str = query[1]
+    if(length(query) > 1) {
+        attr(query_str, "units") <- query[2]
+    }
+
+    return(query_str)
+}
+
+# Design doc:
+# change get/set data to accept a query_args argument that is a dict/list
+# used to substitute {arg_tag@arg_type} tags from the query database with the rules:
+# 1) arg_type = name -> NamedFilter; year -> YearFilter
+# 2) key: None -> +MatchesAny
+# 3) key: [operator, int-ish]
+#    where int-ish is ether an int or '.' support others?
+#    and operators are </>|= or * for any
+#    NOTE: we should support IndexFilter instead of YearFilter but to keep it simple
+#    for the users we will try to detect which they meant by assuming only years will
+#    be four digits long
+# 4) key: [ operator,char]
+#    where operators are either = or =~ for regex or * for any
+# 5) For query tags not specified by the user it will be replaced with nothing:
+# i.e. matches any but collapse
+
+#' Find all placeholder in a raw query and return them in a list.
+#'
+#' Find query placeholders of the form \code{'{tag@type}'} and put them into a
+#' list such as \code{list('tag1' = 'name', 'tag2' = 'year')} for instance.
+#' @param query_str A raw query string.
+#' @return A list of all the placeholders found in the query where the names
+#' are the tags and the values are the types.
+#' @importFrom stringr str_match_all str_split
+find_placeholders <- function(query_str) {
+    raw_placeholders <- str_match_all(query_str, '\\{([\\w@]+)\\}')[[1]][,2]
+    ret <- list()
+    for(placeholder in raw_placeholders) {
+        ph_split = str_split(placeholder, '@')[[1]]
+        if(length(ph_split) != 2) {
+            stop(paste0('Invalid placeholder syntax: ', placeholder, 'expecting two values split by @'))
+        } else if(ph_split[2] != 'name' && ph_split[2] != 'year') {
+            stop(paste0('Invalid placeholder syntax, unknown type: ', ph_split[1], 'expecting name or year'))
+        } else {
+            ret[ph_split[1]] = ph_split[2]
+        }
+    }
+
+    return(ret)
+}
+
+#' Parse user options for integer operators and generate GCAM Fusion syntax
```

Comment:
Yes, I think that is a good idea.  I think in particular we could be more aggressive about default behaviors to minimize the syntax in the use cases that will happen like 90% of the time.

Reply:
In this vain of thought I think I am going to change `get_data` / `set_data` to accept query parameters as `*args` and `**kwargs` instead of a `dict()`.  Thus:
```
params = {‘region’: [‘*’], ‘year’: [‘<=‘, yr]}
gcam.get_data(query, params)
```

Now changes to:
```
gcam.get_data(query, **params)
# which is equivalent to all of the following:
gcam.get_data(query, "region", year = ['<', yr])
gcam.get_data(query, region = None, year = ['<', yr])
gcam.get_data(query, region = [], year = ['<', yr])
gcam.get_data(query, region = ['*'], year = ['<', yr])
```

I think it adds flexibility and the opportunity to simplify syntax in the common use case as `"region"` is going to be used the most by far.

Thoughts?

Reply:
I like this implementation.  With the acceptance of kwargs, we will need to have error checking setup that provides an informative exception if the user gets it wrong.  Optional:  perhaps it would be good to setup a verbose argument with this that echos the query string back out that is used to generate the query so the user can see what is going on if so desired.